File: 238.product-of-array-except-self/solution.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        N = len(nums)
        answers = [0] * N
        total = 1
        zeroCnt = 0
        zeroIdx = -1
        for i, num in enumerate(nums):
            if (num == 0):
                zeroCnt += 1
                zeroIdx = i
            else:
                total *= num
        if (zeroCnt == 1):
            answers[zeroIdx] = total
            
        if (zeroCnt > 0):
            return answers
        
        for i, num in enumerate(nums):
            answers[i] = int(total / num)
        
        return answers

File: 238.product-of-array-except-self/solution.ts
function productExceptSelf(nums: number[]): number[] {
    
  const N = nums.length;
  let zeros = 0;
  
  let total = 1;
  for (let i = 0; i < N; i++) {
      if (nums[i] === 0) {
          zeros++;
          continue;
      }
      total *= nums[i];
  }
  
  switch (zeros) {
      case 0: {
          for (let i = 0; i < N; i++) {
              nums[i] = total / nums[i];
          }
          break;
      }
          
      case 1: {
          for (let i = 0; i < N; i++) {
              if (nums[i] === 0) {
                  nums[i] = total;
                  continue;
              }
              nums[i] = 0;
          }
          break;
      }
      
      default:
          for (let i = 0; i < N; i++) {
              nums[i] = 0;
          }
          break;
  }
  
  return nums;
};